contracts/arbitrage/BancorArbitrage.sol

```diff
@@ -0,0 +1,479 @@
+
+        // perform the trade routes
+        Token sourceToken = Token(address(_bnt));
+        for (uint256 i = 0; i < routes.length; i++) {
+            // save the current balance
+            uint256 previousBalance = routes[i].targetToken.balanceOf(address(this));
+
+            // perform the trade
+            _trade(
+                routes[i].exchangeId,
+                sourceToken,
+                routes[i].targetToken,
+                sourceAmount,
+                routes[i].minTargetAmount,
+                routes[i].deadline,
+                routes[i].customAddress,
+                routes[i].customInt
+            );
+
+            // the current iteration target token is the source token in the next iteration
+            sourceToken = routes[i].targetToken;
+
+            // the resulting trade amount is the source amount in the next iteration
+            sourceAmount = routes[i].targetToken.balanceOf(address(this)) - previousBalance;
+        }
+
+        // return the flashloan
+        erc20Token.safeTransfer(msg.sender, amount + feeAmount);
+    }
+
+    /**
+     * @dev handles the trade logic per route
+     */
+    function _trade(
+        uint256 exchangeId,
+        Token sourceToken,
+        Token targetToken,
+        uint256 sourceAmount,
+        uint256 minTargetAmount,
+        uint256 deadline,
+        address customAddress,
+        uint256 customInt
+    ) private {
+        if (exchangeId == EXCHANGE_ID_BANCOR_V2) {
+            // allow the network to withdraw the source tokens
+            sourceToken.safeApprove(address(_bancorNetworkV2), sourceAmount);
+
+            // build the conversion path
+            address[] memory path = new address[](3);
+            path[0] = address(sourceToken);
+            path[1] = customAddress; // pool token address
+            path[2] = address(targetToken);
+
+            uint256 val = sourceToken.isNative() ? sourceAmount : 0;
+
+            // perform the trade
+            _bancorNetworkV2.convertByPath{ value: val }(
+                path,
+                sourceAmount,
+                minTargetAmount,
+                address(0x0),
+                address(0x0),
+                0
+            );
+
+            return;
+        }
+
+        if (exchangeId == EXCHANGE_ID_BANCOR_V3) {
+            // allow the network to withdraw the source tokens
+            sourceToken.safeApprove(address(_bancorNetworkV3), sourceAmount);
+
+            // perform the trade
+            _bancorNetworkV3.tradeBySourceAmount(
+                sourceToken,
+                targetToken,
+                sourceAmount,
+                minTargetAmount,
+                deadline,
+                address(0x0)
+            );
+
+            return;
+        }
+
+        if (exchangeId == EXCHANGE_ID_UNISWAP_V2 || exchangeId == EXCHANGE_ID_SUSHISWAP) {
+            IUniswapV2Router02 router = exchangeId == EXCHANGE_ID_UNISWAP_V2 ? _uniswapV2Router : _sushiSwapRouter;
```

Comment:
Can probably store the immutable addresses of `_uniswapV2Router.WETH()` and `_sushiSwapRouter.WETH()` upon contract creation, then use them here instead of the external call taking place on every trade.
It would require splitting the code below into two separate cases, one for each router, so probably best to use a common function in order to avoid coding-duplication.

Reply:
already known to be the same WETH for both of them on mainnet so can simply use the already defined state var for that - done.